staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,616 @@
+		return false, false, func() {}
+	case <-ctx.Done():
+		klog.V(5).Infof("request timed out\n")
+		if req.Enqueued {
+			// remove the request from the queue as it has timed out
+			func() {
+				qs.lockAndSyncTime()
+				defer qs.lock.Unlock()
+				for i := range req.Queue.Requests {
+					if req == req.Queue.Requests[i] {
+						req.Queue.Requests = removeRequest(req.Queue.Requests, i)
+						break
+					}
+				}
+			}()
+		} else {
+			qs.counter.Add(-1)
+		}
+		return false, false, func() {}
+	}
+	// ************************************************************************
+}
+
+func (qs *queueSetImpl) getvirtualtimeratio() float64 {
+	NEQ := 0
+	reqs := 0
+	for _, queue := range qs.queues {
+		reqs += queue.RequestsExecuting
+
+		if len(queue.Requests) > 0 || queue.RequestsExecuting > 0 {
+			NEQ++
+		}
+	}
+	// no active flows, virtual time does not advance (also avoid div by 0)
+	if NEQ == 0 {
+		return 0
+	}
+	return math.Min(float64(reqs), float64(qs.concurrencyLimit)) / float64(NEQ)
+}
+
+// synctime is used to sync the time of the queueSetImpl by looking at the elapsed
+// time since the last sync and this value based on the 'virtualtime ratio'
+// which scales inversely to the # of active flows
+func (qs *queueSetImpl) synctime() {
+	realNow := qs.clk.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	qs.vt += timesincelast * qs.getvirtualtimeratio()
+}
+
+func (qs *queueSetImpl) lockAndSyncTime() {
+	qs.lock.Lock()
+	qs.synctime()
+}
+
+// isEmpty is a convenience method that returns 'true' when all of the queues
+// in an queueSetImpl have no packets (and is "empty")
+func (qs *queueSetImpl) isEmpty() bool {
+	return qs.numRequestsEnqueued == 0
+}
+
+// timeoutOldRequestsAndRejectOrEnqueue encapsulates the logic required
+// to validate and enqueue a request for the queueSetImpl/QueueSet:
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a request and enqueue
+// returns the enqueud request on a successful enqueue
+// returns nil in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+// NOTE: This function must only be called with the QueueSet locked
+func (qs *queueSetImpl) timeoutOldRequestsAndRejectOrEnqueue(hashValue uint64) *fq.Request {
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.chooseQueueIdx(hashValue)
+	queue := qs.queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	qs.removeTimedOutRequestsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
```

Comment:
What was obtained that counter balanced the noted detriment?

Reply:
How about documenting the answer by replacing the last line of this comment with:

> // We think this simple approach is good enough, at least for now.

Reply:
This is currently for the simplicity of the approach.  I've updated the comment to read as suggested. Additionally as we have now plumbed timeout handling from the context directly there is an enforceable max timeout which I should make this less of an issue.